### PR TITLE
Add CKEditor as WYSIWYG editor

### LINKED ISSUE #210: Allow Mailbody formating [$100]
### Steps to reproduce
1. Open Mail App
2. Click on Button "New Message"
3. Type some Text in the Textarea for Mailbody

### Expected behaviour
There should be a Buttonbar, which let me format the Mail with the most needed Formats like bold, italic, font color, font size, font name. Maybe its possible to add something like tinymce.

### Actual behaviour
Currently, its a pure Textarea, only normal Text could be entered.

### Mail app

**Mail app version: 0.60**

**Mailserver or service: Personal IMAP Server **

### Server configuration
**Operating system**: Raspbian (Debian, Jessie)

**Web server:** Nginx

**Database:** MariaDB

**PHP version:** 5.6

**Updated from an older version or fresh install:** fresh install

#### Client configuration
**Browser:** Firefox 50

**Operating system:** Windows

<bountysource-plugin>

---
There is a **[$100 open bounty](https://www.bountysource.com/issues/39482772-allow-mailbody-formating?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by ChristophWurst**
Fixes #210 #306.

- [x] Blocked by https://github.com/ckeditor/ckeditor5-vue/issues/94

Todo

- [x] Add CKEdit ([balloon editor](https://ckeditor.com/docs/ckeditor5/latest/examples/builds/balloon-editor.html))
- [x] Adjust design
- [x] Fix conflict with keyboard shortcuts
- [x] keyup handler
- [ ] keypress handler
- [x] Make it possible to disable all rich features (text-only
- [x] L10n https://ckeditor.com/docs/ckeditor5/latest/builds/guides/integration/frameworks/vuejs.html#ready-to-use-builds
- [x] Limit allowed HTML
- [x] Check loading of HTML drafts
- [x] Add an account setting for the default mode

### CODE REVIEW

**Comment by ChristophWurst** [comment]
Still a lot to do, but I really love how nonobstructive this editor is

Bildschirmfoto vom 2019-09-04 11-57-01

**Comment by ChristophWurst** [comment]
Bildschirmfoto von 2019-09-04 15-17-36

**Comment by ChristophWurst** [comment]
While this does has a negative affect on bundle size, I think we're doing still pretty good. Like even compared to our bigger brother Gmail.

### Loading gmail + composing a new message (so their have to load their editor)
Bildschirmfoto von 2019-09-04 15-29-58

### Loading Mail on this branch
Bildschirmfoto von 2019-09-04 15-30-20

**Comment by kesselb** [comment]
Any idea why ckeditor is so huge? I used https://github.com/summernote/summernote a while ago but jquery is a hard dependency. https://github.com/basecamp/trix looks promising. But picking a wysiwyg was always a hard challange :rofl:

**Comment by ChristophWurst** [comment]
Guess it's a mix of a lot of code, many plugins, styles and icons?

Yeah, picking an editor is PITA. But I find this one to be quite okay and also easy to integrate.

**Comment by ChristophWurst** [comment]
I'm sure there are a dozen of edge cases not covered but the basics seem to be done. @nextcloud/mail may I ask you to give this a little test? Would love to hear your feedback :heart:

**Comment by ChristophWurst** [comment]
> I'm sure there are a dozen of edge cases not covered but the basics seem to be done. @nextcloud/mail may I ask you to give this a little test? Would love to hear your feedback heart

:ping_pong: :pray: :sweat_smile:

**Comment by ChristophWurst** [comment]
Hey @matiasdelellis @szaimen have you tested the feature? Does it work for you?

> I'm sure there are a dozen of edge cases not covered but the basics seem to be done. @nextcloud/mail may I ask you to give this a little test? Would love to hear your feedback heart

Hey @nextcloud/mail. If you don't help testing I'll merge and let you suffer from the bugs this change brings. This is the last warning :stuck_out_tongue:

**Comment by matiasdelellis** [comment]
Wow, I didn't try it due I don't have any servers with NC17

I never found out that it became stable.. :man_facepalming: 
(The blog announcement seems just a description of features rather than a new stable version annoucement.. :disappointed: when it should be celebrated!!! :smile: :sweat_smile: :tada: :heart: )

Now I try it again .. :smile:

**Comment by matiasdelellis** [comment]
As a simple user, just test these points and add the comments as subpoints.

 * Write rich mail.
   - imagen
   - Without knowing its existence, it would be difficult to find this option.
   - I don't want to get into the debate of activating it by default or not, but it would be interesting, and relatively easy to make it optional. **[1]**
 
 * Send mail.
   - Ok .. No comments.

 * Receive the reply.
   - Ok .. No comments.

 * Write a reply.
   - Again, the app convert the original rich mail to flat, and continue writing flat.
   - When selecting the rich text option, I lose the formatting of the original text. **[2]**

 * Wait until save as draft.
   - Ok .. No comments.
 * Go to 'draft' folder.
   - Ok... I see the last draft with the rich text...
 * Click on last 'draft'.
   - Shows the plain text again.
 * Now selecting rich text option.
   - Now yes, it shows the original rich text.

In summary, as a simple user, I would like to have point **[1]** as:

 * [Compose HTML message:]
   - [Never]
   - [When replying to HTML messages]
   - [When forwarding or replying to HTML messages]
   - [Always]
   - [Always except when responding plain text]

(These taked from roundclube)

However, the only important thing would be **[2]** not to lose the original text format.

Of course, thank you very much for all the progress you are making ... :smile:

**Comment by ChristophWurst** [comment]
> Without knowing its existence, it would be difficult to find this option

There is also a new option for the default in your account settings. And yes, the goal was to make this non-obtrusive. I'm fine if this is not too obvious at the beginning so we can still fix bugs while many people don't notice the feature :stuck_out_tongue: 

> Again, the app convert the original rich mail to flat, and continue writing flat.

Good point. I didn't test this. Some old code might convert the original HTML message to plain text automatically.

> When selecting the rich text option, I lose the formatting of the original text.

You mean you lose it *as well*?

> Shows the plain text again

Does this also happen if rich text editing is the account default?

> In summary, as a simple user, I would like to have point

That is some good input. And I think I saw something similar when I did some research on how other clients handle it. Could you open a new ticket for that? It's something I wouldn't mind adding in general, but it needs some discussion and isn't a hard requirement for a first working implementation. This logic can follow later on just as well.



> Of course, thank you very much for all the progress you are making

Thank YOU for testing :v:
